docs/concepts/native-interop.rst

```diff
@@ -0,0 +1,254 @@
+"native interoperability" that are available on the .NET platform. 
+
+There are a couple of reasons why you would want to call into native code:
+
+* Operating Systems come with a large volume of APIs that are 
+  not present in the managed class libraries. A prime example for this would be 
+  access to hardware or operating system management functions.
+* Communicating with components that are written in unmanaged languages usually 
+  means using P/Invoke. This covers components written in other managed languages,
+  such as Java for instance.
+* On Windows, most of the software that gets installed, such as Microsoft Office
+  suite, registers COM components that represent their programs and allow developers 
+  to automate them or use them. This also requires native interoperability.
+
+There are certainly more situations where it is apt to use native code, especially 
+those that are specific for a given developer's scenario. 
+
+Platform Invoke (P/Invoke)
+--------------------------
+P/Invoke is a technology that allows you to access structs, callbacks and functions
+in unmanaged libraries from your managed code. Most of the P/Invoke API is contained 
+in two namespaces: ``System`` and ``System.Runtime.InteropServices``. Using these 
+two namespaces will allow you access to the attributes
+
+Let's start from the most common example, and that is reusing unmanaged functions 
+in your managed code. Let's show a message box from a command-line application:
+
+.. code-block:: c#
+    :linenos:
+    
+    using System.Runtime.InteropServices;
+
+    public class Program {
+    
+        [DllImport("user32.dll")]
+        public static extrn int MessageBox(IntPtr hWnd, String text, String caption, int options);
+
+        public static void Main(string[] args) {
+            MessageBox(IntPtr.Zero, "Command-line message box", "Attention!", 0);
+        }
+    }
+
+The example above is pretty simple, but it does show off what is needed to 
+invoke unmanaged functions from managed code. Let's step through the example:
+
+* Line #1 shows the using statement for the ``System.Runtime.InteropServices`` 
+  which is the namespace that holds all of the items we need. 
+* Line #5 introduces the ``DllImport`` attribute. This attribute is crucial, as 
+  it tells the runtime that it should load the unmanaged DLL. This is the DLL 
+  into which we wish to invoke.
+* Line #6 is the crux of the P/Invoke work. It defines a managed method that has
+  the **exact same signature** as the unmanaged one. The declaration has a new 
+  keyword that you can notice, ``extern``, which tells the runtime this is an 
+  external method, and that when you invoke it, the runtime should find it in the 
+  DLL specified in ``DllImport`` attribute.
+
+The rest of the example is just invoking the method as you would any other managed 
+method. 
+
+Of course, the runtime allows communication to flow both ways 
```

Comment:
This might fit better into a more expansive section that describe various interop scenarios.